pkg/stub/handler.go

```diff
@@ -2,96 +2,531 @@ package stub
+	h.secretclientwrapper = &defaultSecretClientWrapper{h: &h}
+
+	h.namespace = getNamespace()
+
+	h.skippedImagestreams = make(map[string]bool)
+	h.skippedTemplates = make(map[string]bool)
+
+	h.mutex = &sync.Mutex{}
+	timer := time.NewTimer(5 * time.Second)
+	go func() {
+		<-timer.C
+		h.CreateDefaultResourceIfNeeded()
+	}()
+
+	return &h
 }
 
 type Handler struct {
-	// Fill me
+	initter InClusterInitter
+
+	sdkwrapper SDKWrapper
+
+	samplesResource *v1alpha1.SamplesResource
+	registrySecret  *corev1.Secret
+
+	restconfig  *restclient.Config
+	tempclient  *templatev1client.TemplateV1Client
+	imageclient *imagev1client.ImageV1Client
+	coreclient  *corev1client.CoreV1Client
+
+	imageclientwrapper    ImageStreamClientWrapper
+	templateclientwrapper TemplateClientWrapper
+	secretclientwrapper   SecretClientWrapper
+
+	fileimagegetter    ImageStreamFromFileGetter
+	filetemplategetter TemplateFromFileGetter
+	filefinder         ResourceFileLister
+
+	namespace string
+
+	skippedTemplates    map[string]bool
+	skippedImagestreams map[string]bool
+
+	mutex *sync.Mutex
 }
 
-func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
-	switch o := event.Object.(type) {
-	case *v1alpha1.SamplesResource:
-		logrus.Printf("ctrl cfg %#v", o)
-		restconfig, err := getRestConfig()
+func (h *Handler) GoodStatusUpdate(srcfg *v1alpha1.SamplesResource, newStatus corev1.ConditionStatus, condition v1alpha1.SamplesResourceConditionType) error {
+	status := srcfg.Condition(condition)
+	// decision was made to not spam master if
+	// duplicate events come it (i.e. status does not
+	// change)
+	if status.Status != newStatus {
+		now := kapis.Now()
+		status.LastUpdateTime = now
+		status.Status = newStatus
+		status.LastTransitionTime = now
+		srcfg.ConditionUpdate(status)
+		err := h.sdkwrapper.Update(srcfg)
 		if err != nil {
-			logrus.Errorf("failed to get rest config : %v", err)
-			return err
+			if !kerrors.IsConflict(err) {
+				return h.processError(srcfg, v1alpha1.SamplesUpdateFailed, err, "failed adding success condition to config: %v")
+			}
+			logrus.Printf("got conflict error %#v on success status update, going retry", err)
+			srcfg, err = h.sdkwrapper.Get(srcfg.Name, srcfg.Namespace)
+			if err != nil {
+				return h.processError(srcfg, v1alpha1.SamplesUpdateFailed, err, "failed to retrieve samples resource after update conflict: %v")
+			}
+			status = srcfg.Condition(condition)
+			status.LastTransitionTime = now
+			status.LastUpdateTime = now
+			status.Status = newStatus
+			srcfg.ConditionUpdate(status)
+			err = h.sdkwrapper.Update(srcfg)
+			if err != nil {
+				// just give up this time
+				return h.processError(srcfg, v1alpha1.SamplesUpdateFailed, err, "failed to update status after conflict retry: %v")
+			}
+		} else {
+			// get the latest copy in case something modified the object underneath us
+			// per observations during testing
+			sr, _ := h.sdkwrapper.Get(srcfg.Name, srcfg.Namespace)
```

Comment:
we just did an update of the object, why would we need to fetch it back? if it does change, we'd get a modified event for it, no?

Reply:
Yeah this stems from the intermittent, repeated conflict scenario, I cited earlier.

I'll try to concisely describe:
1) we have an event come in at version A, but in processing, it  gets a conflict cause etcd has version B (where numerically B > A) ... version B's status conditions timestamps have been modified by something, and are different than version A's ... everything else between the 2 versions same
2) we return error 
3) sdk does requeue / retry based on us returning error 
4) cause of retry, we get the *same* event again, i.e. the same now stale version
5) sdk is hardwired to retry 15 times .. so lots of thrashing as we update the samples content again, since we did not update our samplesresource cache, since we did not successfully update the status conditions

With this change, I have since seen this occur (i.e. item 1 above) at least twice now, and by doing this, have avoided the churn around 2) through 5) 

So, yes, *WHAT* changes the timestamp on the condition ???  I could certainly use any suggestions on how to debug, and sure, we could open a bug somewhere (where exactly?), but with this being intermittent, the reality is unless we can get the precise debug data needed, the bug will sit unresolved.

And this workaround is simple in the interim and is working in my testing.

Reply:
but you're already re-getting the fresh object if you get an update conflict and retrying the update, above, so doesn't that already handle the conflict?

Reply:
After the fact I was worried about my status update possibly updating the version yet again, but yeah, in hindsight that is too paranoid (even for me) .... only spec/objectmeta/metadata updates should result in the version getting updated.

I'll remove that `Get` in the non-error path.

Reply:
well i would expect the status update to change the resource version because the resourceversion just comes from etcd as far as i know and doesn't care what the change was... but you'd still get a new modified event due to that update and then your handler would process it, as normal (ideally process it as a no-op since nothing in the spec changed).

Reply:
That is not what I see in practice ... when thing run without issue
1) I get the initial event
2) process it, update the status
3) i don't get a new version event as a result of the updated status ... the operator is quiet until the relist happens

Other than resource version, and not comparing the specs for deltas before proceding

Reply:
ok, i may not understand how resourceversion is managed by k8s, not going to worry about it for now if things are working as you'd hope.